### PR TITLE
call get_ip only when necessary

### PR DESCRIPTION
**Pull request by Tianhao-Gu**
Ensure that `get_ip()` is not called if the environment variable CALLBACK_IP is already set, as invoking `get_ip()` may lead to errors in environments (Argonne sequoia) with restricted outbound traffic.

### CODE REVIEW

File: JobRunner/Callback.py [hunk 1]
```diff
@@ -12,7 +12,9 @@ class Callback():
     def __init__(self):
         workdir = os.environ.get("JOB_DIR", '/tmp/')
         self.conf = Config(job_id="callback", workdir=workdir, use_ee2=False)
-        self.ip = os.environ.get('CALLBACK_IP', get_ip())
+        self.ip = os.environ.get('CALLBACK_IP')
```
Comment (bio-boris) [inline JobRunner/Callback.py:15]: Would `self.ip = os.environ.get('CALLBACK_IP') or get_ip()` work?

**Comment by bio-boris** [comment]
Hi! What  is the context for this change?  Can you please link update the description please? Are there any testing instructions for this change? I think we should set up coverage as well and ensure that future changes are unit tested. Thanks!

File: JobRunner/config.py [hunk 2]
```diff
@@ -34,7 +35,7 @@ def _get_admin_token():
 class Config:
     def __init__(self, workdir=None, base_url=None, job_id=None, use_ee2=True):
         self.job_id = job_id
-        self.base_url = "https://ci.kbase.us/services/"
+        self.base_url = os.environ.get(_KB_BASE_URL, "https://ci.kbase.us/services/")
```
Comment (bio-boris) [inline JobRunner/config.py:38]: No action required: I put this into chatgpt and this is what it gave on how to refactor this code
```
import logging
import os
from socket import gethostname

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class Config:
    DEFAULT_BASE_URL = "https://ci.kbase.us/services/"
    DEFAULT_MAX_TASKS = 10
    ENV_TOKEN = "KB_AUTH_TOKEN"
    ENV_ADMIN_TOKEN = "KB_ADMIN_AUTH_TOKEN"
    ENV_BASE_URL = "KB_BASE_URL"
    ENV_DEBUG = "JOBRUNNER_DEBUG_MODE"
    ENV_MAX_TASKS = "JR_MAX_TASKS"
    ENV_RUNTIME = "RUNTIME"

    def __init__(self, workdir=None, base_url=None, job_id=None, use_ee2=True):
        self.job_id = job_id
        self.base_url = os.getenv(self.ENV_BASE_URL, self.DEFAULT_BASE_URL)
        self.workdir = workdir or os.getcwd()
        self.debug = os.getenv(self.ENV_DEBUG, "").lower() == "true"
        self.runtime = os.getenv(self.ENV_RUNTIME, "docker")
        self.max_tasks = int(os.getenv(self.ENV_MAX_TASKS, self.DEFAULT_MAX_TASKS))
        self.token = self._get_token()
        self.admin_token = self._get_admin_token()
        self.hostname = gethostname()
        self._setup_urls(use_ee2)
        os.makedirs(self.workdir, exist_ok=True)

    def _get_token(self, env_var=ENV_TOKEN, fallback_file="token"):
        token = os.getenv(env_var)
        if not token and os.path.exists(fallback_file):
            with open(fallback_file) as file:
                token = file.read().strip()
                os.environ[env_var] = token
        return token

    def _get_admin_token(self):
        token = os.getenv(self.ENV_ADMIN_TOKEN)
        if token:
            os.environ.pop(self.ENV_ADMIN_TOKEN, None)  # Attempt to remove after retrieval
            if self.ENV_ADMIN_TOKEN in os.environ:
                logger.error("Failed to sanitize environment: admin token still present")
        return token

    def _setup_urls(self, use_ee2):
        auth_ext = "auth/api/legacy/KBase/Sessions/Login"
        self.ee2_url = f"{self.base_url}ee2" if use_ee2 else None
        self.catalog_url = f"{self.base_url}catalog"
        self.auth_url = f"{self.base_url}{auth_ext}"
        self.auth2_url = f"{self.base_url}auth/api/V2/token"
```

File: JobRunner/Callback.py (hunk 1 above)
Comment (Tianhao-Gu) [reply JobRunner/Callback.py]: 👍 sure. It works exactly the same way. More cleaner.

**Comment by Tianhao-Gu** [comment]
> Hi! What is the context for this change? Can you please link update the description please? Are there any testing instructions for this change? I think we should set up coverage as well and ensure that future changes are unit tested. Thanks!

👍 added tests and desc.

File: JobRunner/config.py (hunk 2 above)
Comment (Tianhao-Gu) [reply JobRunner/config.py:38]: The suggested approach appears much cleaner, but for now, I guess we can just stick with the current working code.

**Comment by bio-boris** [comment]
Thanks! Glad to hear that these latest changes work on Sequoia.

Can you please check out the tests and see what is going on with them? I believe it says that they are running, but if you look at the logs, it looks like they are failing. Are they indeed failing, or are the tests expected to fail that way?

**Comment by Tianhao-Gu** [comment]
> Thanks! Glad to hear that these latest changes work on Sequoia.
> 
> Can you please check out the tests and see what is going on with them? I believe it says that they are running, but if you look at the logs, it looks like they are failing. Are they indeed failing, or are the tests expected to fail that way?

All tests passed. Some tests might just test failures. I reviewed some past GHA test runs, and the test logs are structured like this.

```
Ran 33 tests in 42.356s

OK
```

**Review by bio-boris — APPROVED** [review]
LGTM!

**Pull request merged**